Updated Backup section (#4103)
gravitational/teleport: docs/4.3/admin-guide.md

    to SHA-512 based signatures. To upgrade an existing cluster, set the following in
    your teleport.yaml:

    ```bash
    teleport:

>       ca_signature_algo: “rsa-sha2-512”

Reviewer: ```suggestion
      ca_signature_algo: "rsa-sha2-512"
```

Change the smart quotes -> regular quotes so they can be copy/pasted.

## Backing Up Teleport

When planning a backup of Teleport, it's important to know what is where and the

> importance of each component. Teleports Proxies and Nodes are stateless, and thus

Reviewer 2: ```suggestion
importance of each component. Teleport's Proxies and Nodes are stateless, and thus
```

only `teleport.yaml` should be backed up.

> The Auth server is Teleports brains, and depending on the backend should be backup

Reviewer 2: ```suggestion
The Auth server is Teleport's brains, and depending on the backend should be backed up
```

in a regular cadance.

> For example a customer running Teleport on AWS with etd have these key items of data:

Reviewer 3: `s/etd/DynamoDB/`

> | What | Where ( Example AWS Customer ) |

Reviewer 3: Also mention audit log and session recordings

|-|-|
| Local Users ( not SSO )  | DynamoDB |
| Certificate Authorities | DynamoDB |
| Trusted Clusters | DynamoDB |
| Connectors: SSO | DynamoDB / File System  |
| RBAC | DynamoDB / File System |
| teleport.yaml | File System |
| teleport.service | File System |
| license.pem | File System |

> | TLS /. | ( File System / Outside Scope )  |

Reviewer 3: `TLS /.` -> `TLS key/certificate`

For this customer, we would Recommend using [AWS best practices](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/BackupRestore.html) for backing up DynamoDB. If DynamoDB is used for
the audit log, logged events have a TTL of 1 year.

> | Backend | Recommend Backup Strategy  |

Reviewer 3: nit: Why Capital Case Everywhere?

|-|-|

> | dir ( local filesystem )   | Copy `data_dir/storage` and use `tctl get all` to get local state. |

Reviewer 3: I'd rephrase the strategy as: "Backup `/var/lib/teleport/storage` directory and the output of `tctl get all`".
Also, if a user backs up `/var/lib/teleport`, is there any reason to do `tctl get all`?

Reviewer: I think the idea with `tctl get all` was that the format would be YAML and therefore backend agnostic, so theoretically you could dump out the CAs, users etc and then restore these into a different backend.

| DynamoDB | [Follow AWS Guidelines for Backup & Restore](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/BackupRestore.html) |

> | etcD | [Follow etcD Guidleines for Disaster Recovery ](https://etcd.io/docs/v2/admin_guide) |

Reviewer 3: nit: etcd is all lowercase

| Firestore | [Follow GCP Guidlines for Automated Backups](https://firebase.google.com/docs/database/backups) |

### Config Files

> We would recommend you store configs as discrete files in git.

Reviewer 3: Elaborate that this refers to roles, trusted clusters, local users and auth connectors. Maybe we can invent a better term for these :thinking: 
This does not refer to `teleport.yaml` itself.

Author: We call them Resources. I've updated the doc

> We would recommend you store configs as discrete files in git.

Reviewer 3: `git` -> `a version control system`

### GitOps

> If running Teleport at scale it's important for teams to have an automated way to

Reviewer 2: ```suggestion
If running Teleport at scale, it's important for teams to have an automated way to
```

> restore Teleport. At a high level, this is our recommend approach:

Reviewer 3: I wouldn't classify this as backup/restore.
This is more about applying configs by committing files to a repo.

- Persist and backup your backend
- Share that backend among auth servers

> - Store your configs as discrete files in git

Reviewer 3: `git` -> `VCS`

> - Have your CI run `tctl create -f *.yaml` from that git directory, without restarting teleport service / pods in k8s

Reviewer 3: skip the ", without restarting teleport service / pods in k8s" and mention that `tctl create` must run on the auth server

> - If you lose your backend, this setup still restores the essential configs

Reviewer 3: no need to include this last bullet

## Backing Up Teleport

When planning a backup of Teleport, it's important to know what is where and the
importance of each component. Teleports Proxies and Nodes are stateless, and thus
only `teleport.yaml` should be backed up.

The Auth server is Teleports brains, and depending on the backend should be backup
in a regular cadance.

For example a customer running Teleport on AWS with DynamoDB have these key items of data:

| What | Where ( Example AWS Customer ) |
|-|-|
| Local Users ( not SSO )  | DynamoDB |
| Certificate Authorities | DynamoDB |
| Trusted Clusters | DynamoDB |
| Connectors: SSO | DynamoDB / File System  |
| RBAC | DynamoDB / File System |
| teleport.yaml | File System |
| teleport.service | File System |
| license.pem | File System |
| TLS key/certificate | ( File System / Outside Scope )  |
| Audit log | DynamoDB  |
| Session recordings| S3  |

For this customer, we would recommend using [AWS best practices](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/BackupRestore.html) for backing up DynamoDB. If DynamoDB is used for
the audit log, logged events have a TTL of 1 year.

> | Backend | Recommend backup strategy  |

Reviewer: ```suggestion
| Backend | Recommended backup strategy  |
```

|-|-|
| dir ( local filesystem )   | Copy `data_dir/storage` and use `tctl get all` to get local state. |
| DynamoDB | [Follow AWS Guidelines for Backup & Restore](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/BackupRestore.html) |
| etcd | [Follow etcD Guidleines for Disaster Recovery ](https://etcd.io/docs/v2/admin_guide) |
| Firestore | [Follow GCP Guidlines for Automated Backups](https://firebase.google.com/docs/database/backups) |

### Config Files

We would recommend you store configs as discrete files in git.

### GitOps

If running Teleport at scale it's important for teams to have an automated way to

> restore Teleport. At a high level, this is our recommend approach:

Reviewer: ```suggestion
restore Teleport. At a high level, this is our recommended approach:
```

- Persist and backup your backend
- Share that backend among auth servers
- Store your configs as discrete files in git
- Have your CI run `tctl create -f *.yaml` from that git directory, without restarting teleport service / pods in k8s

## Migrating Backends.
As of version v4.1 you can now quickly export a collection of resources from
Teleport. This feature was designed to help customers migrate from local storage
to etcd.

> Using `tctl get all` will retrive the below items:

Reviewer 2: ```suggestion
Using `tctl get all` will retrieve the below items:
```

## Backing Up Teleport

When planning a backup of Teleport, it's important to know what is where and the
importance of each component. Teleport's Proxies and Nodes are stateless, and thus
only `teleport.yaml` should be backed up.

The Auth server is Teleport's brains, and depending on the backend should be backed up

> in a regular cadance.

Reviewer: ```suggestion
regularly.
```

(can't decide on "at a regular cadence" versus "in a regular cadence", so let's just change it to "regularly"?)

    - OIDC [Teleport Enterprise]
    - Roles [Teleport Enterprise]

> When migrating backend you'll should backup up your auth server's `data_dir/storage` directly.

Reviewer 3: `you'll` -> `you`

    - OIDC [Teleport Enterprise]
    - Roles [Teleport Enterprise]

> When migrating backend you should backup up your auth server's `data_dir/storage` directly.

Reviewer: ```suggestion
When migrating backends, you should back up your auth server's `data_dir/storage` directly.
```
